Python/optimizer_analysis.c

```diff
@@ -537,6 +571,99 @@ peephole_opt(_PyInterpreterFrame *frame, _PyUOpInstruction *buffer, int buffer_s
     }
 }
 
+
+static bool
+function_decide_simple_inlineable(
+    _PyUOpInstruction *func_body_start,
+    _PyUOpInstruction *func_body_end)
+{
+    // Usually means MAKE_CELL or something
+    if (func_body_start->opcode != _RESUME_CHECK) {
+        return false;
+    }
+    func_body_start++;
+    _PyUOpInstruction *curr = func_body_start;
+    while (curr < func_body_end) {
+        int opcode = curr->opcode;
+        // We should be the leaf frame.
+        assert(opcode != _PUSH_FRAME && opcode != _PUSH_FRAME_INLINEABLE);
+        if (opcode == _POP_FRAME) {
+            return true;
+        }
+        if (_PyUop_Flags[curr->opcode] & (HAS_ESCAPES_FLAG | HAS_DEOPT_FLAG | HAS_ERROR_FLAG)) {
+            // Pure overrides error flag.
+            if (!(_PyUop_Flags[curr->opcode] & HAS_PURE_FLAG)) {
+                return false;
+            }
+        }
+        curr++;
+    }
+    Py_UNREACHABLE();
+}
+
+static void
+inline_simple_frames(_PyUOpInstruction *buffer, int buffer_size)
+{
+    bool did_inline = false;
+    for (int pc = 0; pc < buffer_size; pc++) {
+        int opcode = buffer[pc].opcode;
+        switch (opcode) {
+            case _PUSH_FRAME_INLINEABLE: {
+                assert(buffer[pc - 3].opcode == _CHECK_STACK_SPACE);
+                assert(buffer[pc - 2].opcode == _INIT_CALL_PY_EXACT_ARGS);
+                assert(buffer[pc - 1].opcode == _SAVE_RETURN_OFFSET);
+                assert(buffer[pc + 1].opcode == _CHECK_VALIDITY_AND_SET_IP ||
+                    buffer[pc + 1].opcode == _CHECK_VALIDITY);
+                // Skip over the CHECK_VALIDITY when deciding,
+                // as those can be optimized away later.
+                if (!function_decide_simple_inlineable(&buffer[pc + 2], buffer + buffer_size)) {
+                    buffer[pc].opcode = _PUSH_FRAME;
+                    break;
+                }
+                assert(buffer[pc + 2].opcode == _RESUME_CHECK);
+                did_inline = true;
+                uint64_t operand = buffer[pc].operand;
+                int locals_len = (int)(operand >> 32);
+                int stack_len = (int)(operand & 0xFFFFFFFF);
+                REPLACE_OP(&buffer[pc - 3], _GROW_TIER2_FRAME, locals_len + stack_len, 0);
+                REPLACE_OP(&buffer[pc - 2], _NOP, 0, 0);
+                REPLACE_OP(&buffer[pc - 1], _NOP, 0, 0);
+                REPLACE_OP(&buffer[pc], _PRE_INLINE, locals_len, 0);
+                REPLACE_OP(&buffer[pc + 1], _NOP, 0, 0);
+                REPLACE_OP(&buffer[pc + 2], _NOP, 0, 0);
+                break;
+            }
+            case _POP_FRAME: {
+                if (did_inline) {
+                    buffer[pc].oparg = (int)buffer[pc].operand;
+                    buffer[pc].opcode = _POST_INLINE;
```

Comment:
Also a good `REPLACE_OP()` candidate. (Looks like I think of `REPLACE_OP()` as changing the opcode field. :-)